[RISCV] Improve legalization of e8 m8 VL>256 shuffles

preames opened the pull request:
If we can't produce a large enough index vector in i8, we may need to legalize
the shuffle (via scalarization - which in turn gets lowered into stack usage).
This change makes two related changes:
* Deferring legalization until we actually need to generate the vrgather
  instruction.  With the new recursive structure, this only happens when
  doing the fallback for one of the arms.
* Check the actual mask values for something outside of the representable
  range.

Both are covered by recently added tests.

topperc commented:
The description here is misleading. I believe returning SDValue() here causes the shuffle to get expanded through memory. It will go through the Expand path in LegalizeDAG.

topperc commented:
> The description here is misleading. I believe returning SDValue() here causes the shuffle to get expanded through memory. It will go through the Expand path in LegalizeDAG.

It doesn't go through memory. I guess I was thinking of build_vector. I'm not sure what shuffle expansion does. I guess it uses extracts and_vector?

preames committed:
[RISCV] Improve legalization of e8 m8 VL>256 shuffles

If we can't produce a large enough index vector in i8, we may need to legalize
the shuffle (via scalarization - which in turn gets lowered into stack usage).
This change makes two related changes:
* Defering legalization until we actually need to generate the vrgather
  instruction.  With the new recursive structure, this only happens when
  doing the fallback for one of the arms.
* Check the actual mask values for something outside of the representable
  range.

Both are covered by recently added tests.

preames commented:
I went ahead and added test coverage for both the original code and the new case.  In the process, I discovered that one my earlier refactorings actually broke this logic - I think my mental state was assuming the prior version of this patch despite it not having landed - and I landed a separate change to get us back to a working baseline.  

Pushed a reworked change, please take a look at the description from scratch.

lukel97 reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -4936,8 +4936,11 @@ static SDValue lowerVECTOR_SHUFFLE(SDValue Op, SelectionDAG &DAG,
-    if (VT.getScalarSizeInBits() == 8 && VT.getVectorNumElements() > 256) {
-      // On such a large vector we're unable to use i8 as the index type.
+    if (VT.getScalarSizeInBits() == 8 &&
+        any_of(enumerate(Mask), [&](const auto &Idx) {
```
Why do we need enumerate?

preames committed:
Address style comment in review
